File: src/lib/scan-store.ts
type ScanStatus = "queued" | "scanning" | "completed" | "failed";

export type PersistedFinding = {
  packageName: string;
  ecosystem: string;
  vulnerabilityId: string;
  cveId?: string | null;
  kevStatus: boolean;
  severity: string;
  summary: string;
  fix: string;
  affectedVersions?: string | null;
  advisoryUrl?: string | null;
  epssScore?: number | null;
  isDirect?: boolean;
};

export type ScanJob = {
  id: string;
  repositoryUrl: string;
  owner: string;
  repo: string;
  branch?: string | null;
  commitSha?: string | null;
  status: ScanStatus;
  createdAt: string;
};

const jobs = new Map<string, ScanJob>();
const findingsByJob = new Map<string, PersistedFinding[]>();
/** commitSha → scanJobId for cache hits */
const shaIndex = new Map<string, string>();

function makeId() {
  return (
    globalThis.crypto?.randomUUID?.() ??
    `${Date.now()}-${Math.random().toString(36).slice(2, 10)}`
  );
}

export async function createScanJob(input: {
  repositoryUrl: string;
  owner: string;
  repo: string;
  branch?: string | null;
  commitSha?: string | null;
}) {
  const job: ScanJob = {
    id: makeId(),
    repositoryUrl: input.repositoryUrl,
    owner: input.owner,
    repo: input.repo,
    branch: input.branch ?? null,
    commitSha: input.commitSha ?? null,
    status: "scanning",
    createdAt: new Date().toISOString(),
  };
  jobs.set(job.id, job);
  findingsByJob.set(job.id, []);
  return job;
}

export async function completeScanJob(scanJobId: string) {
  const job = jobs.get(scanJobId);
  if (job) {
    job.status = "completed";
    if (job.commitSha) shaIndex.set(job.commitSha, job.id);
  }
}

export async function failScanJob(scanJobId: string) {
  const job = jobs.get(scanJobId);
  if (job) job.status = "failed";
}

export async function saveScanFindings(
  scanJobId: string,
  findings: PersistedFinding[],
) {
  if (findings.length === 0) return;
  const existing = findingsByJob.get(scanJobId) ?? [];
  findingsByJob.set(scanJobId, [...existing, ...findings.map((f) => ({ ...f }))]);
}

export async function listRecentScans(limit = 10) {
  const sortedJobs = Array.from(jobs.values()).sort((a, b) =>
    a.createdAt < b.createdAt ? 1 : -1,
  );
  return sortedJobs.slice(0, limit).map((job) => {
    const jobFindings = findingsByJob.get(job.id) ?? [];
    const vulnerableDependencies = new Set(jobFindings.map((f) => f.packageName));
    const kevCount = jobFindings.filter((f) => f.kevStatus).length;
    return {
      ...job,
      findingsCount: jobFindings.length,
      vulnerableDependencies: vulnerableDependencies.size,
      kevCount,
    };
  });
}

export async function getScanById(scanId: string) {
  const job = jobs.get(scanId);
  if (!job) return null;
  const findings = findingsByJob.get(scanId) ?? [];
  return { job, findings };
}

/** Returns a completed scan for the same commit SHA, or null (cache miss). */
export async function findScanByCommitSha(commitSha: string) {
  const jobId = shaIndex.get(commitSha);
  if (!jobId) return null;
  return getScanById(jobId);
}

File: backend/app/scan_store.py
"""In-memory scan store — identical logic to the original TypeScript scan-store.ts."""
from __future__ import annotations

import uuid as _uuid
from datetime import datetime, timezone
from typing import Dict, List, Optional

from app.models import PersistedFinding, RecentScan, ScanDetailResponse, ScanJob

# ── Storage ───────────────────────────────────────────────────────────────────

_jobs: Dict[str, ScanJob] = {}
_findings_by_job: Dict[str, List[PersistedFinding]] = {}
_sha_index: Dict[str, str] = {}  # commitSha → scanJobId


def _make_id() -> str:
    return str(_uuid.uuid4())


# ── CRUD ──────────────────────────────────────────────────────────────────────

def create_scan_job(
    repository_url: str,
    owner: str,
    repo: str,
    branch: Optional[str] = None,
    commit_sha: Optional[str] = None,
) -> ScanJob:
    job = ScanJob(
        id=_make_id(),
        repository_url=repository_url,
        owner=owner,
        repo=repo,
        branch=branch,
        commit_sha=commit_sha,
        status="scanning",
        created_at=datetime.now(timezone.utc).isoformat(),
    )
    _jobs[job.id] = job
    _findings_by_job[job.id] = []
    return job


def complete_scan_job(scan_job_id: str) -> None:
    job = _jobs.get(scan_job_id)
    if job:
        job.status = "completed"
        if job.commit_sha:
            _sha_index[job.commit_sha] = job.id


def fail_scan_job(scan_job_id: str) -> None:
    job = _jobs.get(scan_job_id)
    if job:
        job.status = "failed"


def save_scan_findings(scan_job_id: str, findings: List[PersistedFinding]) -> None:
    if not findings:
        return
    existing = _findings_by_job.get(scan_job_id, [])
    _findings_by_job[scan_job_id] = existing + findings


def list_recent_scans(limit: int = 10) -> List[RecentScan]:
    sorted_jobs = sorted(_jobs.values(), key=lambda j: j.created_at, reverse=True)
    result: List[RecentScan] = []
    for job in sorted_jobs[:limit]:
        job_findings = _findings_by_job.get(job.id, [])
        vulnerable_deps = len({f.package_name for f in job_findings})
        kev_count = sum(1 for f in job_findings if f.kev_status)
        result.append(
            RecentScan(
                id=job.id,
                repository_url=job.repository_url,
                owner=job.owner,
                repo=job.repo,
                commit_sha=job.commit_sha,
                status=job.status,
                created_at=job.created_at,
                findings_count=len(job_findings),
                vulnerable_dependencies=vulnerable_deps,
                kev_count=kev_count,
            )
        )
    return result


def get_scan_by_id(scan_id: str) -> Optional[ScanDetailResponse]:
    job = _jobs.get(scan_id)
    if not job:
        return None
    findings = _findings_by_job.get(scan_id, [])
    return ScanDetailResponse(job=job, findings=findings)


def find_scan_by_commit_sha(commit_sha: str) -> Optional[ScanDetailResponse]:
    job_id = _sha_index.get(commit_sha)
    if not job_id:
        return None
    return get_scan_by_id(job_id)
